proto/profile.proto

```diff
@@ -0,0 +1,109 @@
+syntax = "proto3";
+
+package io.linkerd.proxy.profile;
+
+option go_package = "github.com/linkerd/linkerd2-proxy-api/go/profile";
+
+import "google/protobuf/duration.proto";
+
+import "http_types.proto";
+import "destination.proto";
+
+service Profile {
+  // Given a destination, return that service's profile and send an update
+  // whenever it changes.
+  rpc Get(destination.GetDestination) returns (stream ServiceProfile) {}
+}
+
+message ServiceProfile {
+  // A list of request polcies, each with a RequestMatch.  If a request matches
+  // more than one RequestPolicy, the first match wins.
+  repeated RequestPolicy request_policies = 1;
+  // Retry budget describes how much additional load is acceptable to send to
+  // this service from retries.
+  RetryBudget retry_budget = 2;
+  google.protobuf.Duration default_retry_timeout = 3;
+}
+
+message RetryBudget {
+  // The maximum ratio of retries to non-retry requests.  For example if this
+  // value is 0.2 then the number of retries is at most 20% of the number of
+  // non-retry requests.  This can be thought of as the amount of extra load
+  // added by retries (e.g. 20% extra load).
+  float max_retry_ratio = 1;
+  // The duration over which the retry budget is calculated.  For example, if
+  // this value is 10 seconds, then in the past 10 seconds it must be true that:
+  //
+  // #retries <= #requests*max_retry_ratio + additional_retries_per_second*10
+  //
+  // Increasing the duration over which this is calculated allows for larger
+  // bursts of retries.
+  google.protobuf.Duration ttl = 2;
+  // This amount of retries per second are allowed in addition to the ones
+  // allowed by the max_retry_ratio.  This is to allow retries to happen even
+  // when the request rate is low.
+  uint32 additional_retries_per_second = 3;
+}
+
+message RequestPolicy {
+  // This policy applies to requests that match this condition.
+  RequestMatch condition = 1;
+  repeated ResponsePolicy response_policies = 2;
+  // After this duration has elapsed since the initial request, no more retries
+  // will be attempted.  If unspecified, this defaults to the service profile's
+  // default_retry_timeout.
+  google.protobuf.Duration retry_timeout = 3;
+}
+
+message ResponsePolicy {
+  // This policy applies to responses that match this condition.
+  ResponseMatch condition = 1;
+  string description = 2;
+  // If responses that match this policy should be considered failures.
+  bool failure = 3;
+  // If responses that match this policy are safe to retry.
+  bool retryably = 4;
```

Comment:
per conversation: i think we should omit description and failure for now and introduce them when the use case is more concrete. For instance, I think a longer _description_ may be useful for logging, but we may want to have more constrained labeling for metrics export, and we may want an arbitrary map in that case (or a map that allows us to extract message metadata into labels).

I also think _retryably_ should be called _is_retryable_ (makes flag-ness clear)

Reply:
Even though failure will always be set to true for now, I'd like to keep it because it really helps illustrate what's going on here.